### DISCUSSION TITLE
Apple callback url ignored? Google and email working fine

### DISCUSSION THREAD

**Discussion by simonbarker** [nextauthjs/next-auth#5878] [Help]
### Question 💬

I have an app deployed to staging that I can sign in to with Apple however rather than redirecting back to the correct callback url (added in the signIn method as a second param) just goes to the domain.

The app is hosted at a path so it should go to subdomain.domain.com/myApp with some query params but instead just goes to subdomain.domain.com without the query params. 

I can see the url is passed into to 

```
callbacks: {
      async redirect({ url, baseUrl }) {
```

correctly and I've even overridden the return to pass back the url variable no matter what but still no dice.

There are no errors in the logs, there is a valid session because if I manually add the path and query params to the url then it works as expected. 

The path and params just aren't being passed around correctly I guess?

This all works totally fine with Google sign in and email magic link sign in so really feels like it might be a bug in the apple provider OR the way I've set up apple (but it all looks fine in the dev portal)

Any suggestions would be great as I've lost a lot of hours to this and have no idea where to go next.

### How to reproduce ☕️

This should work for Apple

```
return signIn("google", {
      callbackUrl: `${process.env.NEXT_PUBLIC_APP_URL}/?id=${productId}&noScan=true`,
    });
```
And it doesn't, instead it just goes to the url in `process.env.NEXT_PUBLIC_APP_URL` which has the pattern:

`https://subdomain-dev.domain.tt/appPath`

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

### npx envinfo --system --binaries --browsers --npmPackages "next,react,next-auth"

System:
    OS: Linux 5.4 Debian GNU/Linux 10 (buster) 10 (buster)
    CPU: (1) x64 Intel(R) Xeon(R) CPU @ 2.30GHz
    Memory: 2.87 GB / 3.60 GB
    Container: Yes
    Shell: 5.0.3 - /bin/bash
  Binaries:
    Node: 16.14.2 - /usr/local/bin/node
    Yarn: 1.22.18 - /usr/local/bin/yarn
    npm: 8.5.0 - /usr/local/bin/npm
  npmPackages:
    next: 12.3.1 => 12.3.1 
    next-auth: ^4.15.0 => 4.15.0 
    react: 18.2.0 => 18.2.0

**Comment by nmandal** [comment]
I have been struggling for weeks implementing Sign In with Apple.

Have tried [this post](https://www.kyle-melton.com/blog/2022-02-how-to-setup-sign-in-with-apple), [this post](https://www.thesiddd.com/blog/apple-auth), as well as many GH and StackOverflow threads to no avail.

Happy to help where I can - would also be interested in where you're at in the process.

**Comment by nmandal** [comment]
Also tagging this [discussion](https://github.com/nextauthjs/next-auth/discussions/3503#top) from NextAuth discussions.

**Comment by simonbarker** [comment]
Thanks @nmandal I'd read the first but not the second of those posts. Neither have anything different in them from what I've implemented sadly.

I wasn't able to get anything working at all on local host, once we deployed to staging with a real url it was much easier to actually get the auth process working - just, like I say without the final redirect working correctly.

Also to create the signed jwt just used the script linked from the next auth apple provider page https://gist.github.com/balazsorban44/09613175e7b37ec03f676dcefb7be5eb

Hope this helps you and if anyone else knows why my callback url isn't being honoured by NextAuth with Apple please throw in suggestions 😀

**Comment by nmandal** [comment]
Tagging the author of the first post for visibility @kalm42

**Comment by dlazares** [comment]
@nmandal the fix in here worked for me https://github.com/nextauthjs/next-auth/issues/2124

I used this to generate my JWT
https://gist.github.com/dlazares/c68fec4b0fa05a631a5452e8a050cd57

my `pages/api/auth/[...nextAuth].ts` looks like this:

```
import { NextApiHandler } from "next";
import NextAuth from "next-auth";
import { PrismaAdapter } from "@next-auth/prisma-adapter";
import GitHubProvider from "next-auth/providers/github";
import prisma from "../../../lib/prisma";
import AppleProvider from "next-auth/providers/apple";

const authHandler: NextApiHandler = (req, res) => NextAuth(req, res, options);
export default authHandler;

const options = {
  cookies: {
    callbackUrl: {
      name: `__Secure-next-auth.callback-url`,
      options: {
        httpOnly: false,
        sameSite: "none",
        path: "/",
        secure: true,
      },
    },
  },
  providers: [
    GitHubProvider({
      clientId: process.env.GITHUB_ID,
      clientSecret: process.env.GITHUB_SECRET,
    }),
    AppleProvider({
      authorization: {
        params: {
          scope: "name email",
          response_mode: "form_post",
          response_type: "code",
        },
      },
      clientId: process.env.APPLE_CLIENT_ID,
      clientSecret: process.env.APPLE_CLIENT_SECRET,
    }),
  ],
  adapter: PrismaAdapter(prisma),
  secret: process.env.SECRET,
};
```

**Comment by simonbarker** [comment]
@dlazares Thanks for this link and example - works perfectly - so weird how it works without this for Google and not for Apple 🤷‍♂️

**Comment by kohj1018** [comment]
Any additional workarounds for this issue?

The problem seems to be due to a cookie named "__Secure-next-auth.callback-url" .

```
cookies: {
    callbackUrl: {
      name: `__Secure-next-auth.callback-url`,
      options: {
        httpOnly: false,
        sameSite: "none",
        path: "/",
        secure: true,
      },
    },
  },
```

The above solution obviously works, but if you write it as above, the callback url of Google sign in does not work.

Therefore, it is not possible to implement Google sign in and Apple sign in at the same time.

**Comment by kohj1018** [comment]
@kohj1018 
I just found a solution to the problem I mentioned.
When the sameSite value is "None", the callbackUrl of Google sign in does not work.

This can be fixed by changing the default configuration of the callbacks' redirect function.
```typescript
import NextAuth from 'next-auth'
import GoogleProvider from 'next-auth/providers/google'
import AppleProvider from 'next-auth/providers/apple'

export default NextAuth({
  cookies: {
    callbackUrl: {
      name: `__Secure-next-auth.callback-url`,
      options: {
        httpOnly: true,
        sameSite: "None",
        path: "/",
        secure: true,
      },
    },
  },
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID as string,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET as string,
      authorization: {
        params: {
          prompt: "consent",
          access_type: "offline",
          response_type: "code"
        }
      }
    }),
    AppleProvider({
      clientId: process.env.APPLE_CLIENT_ID as string,
      clientSecret: process.env.APPLE_CLIENT_SECRET as string,
      authorization: {
        params: {
          scope: "name",
          response_mode: "form_post",
          response_type: "code",
        },
      },
    })
  ],
  secret: process.env.JWT_SECRET,
  callbacks: {
    async redirect({ url, baseUrl }) {

      // Allows relative callback URLs
      if (url.startsWith("/")) return `${baseUrl}${url}`

     // TODO: Below is how redirect works by default. Comment it out and fix it.
      // // Allows callback URLs on the same origin
      // else if (new URL(url).origin === baseUrl) return url

      // return baseUrl

      return url  // Just pass the url
    },
  }
})
```

I hope this solution will help those struggling with the same problem.
